test/TimeLock.js

```diff
@@ -0,0 +1,180 @@
+
+    await expect(
+      timeLock.scheduleSetNum(numHash, eta),
+    ).to.be.revertedWithCustomError(timeLock, 'TimeLockAlreadyScheduled');
+  });
+
+  it('should revert when cancelling a timelock with invalid hash', async () => {
+    await expect(timeLock.cancelSetNum(HashZero)).to.be.revertedWithCustomError(
+      timeLock,
+      'InvalidTimeLockHash',
+    );
+  });
+
+  it('should cancel a timelock', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+    expect(etaTimeLock).to.equal(eta);
+
+    await timeLock.cancelSetNum(numHash).then((tx) => tx.wait());
+
+    const cancelledEta = await timeLock.getTimeLock(numHash);
+    expect(cancelledEta).to.equal(0);
+  });
+
+  it('should revert on executing a timelock with invalid hash', async () => {
+    const num = 5;
+
+    await expect(timeLock.setNum(HashZero, num)).to.be.revertedWithCustomError(
+      timeLock,
+      'InvalidTimeLockHash',
+    );
+  });
+
+  it('should revert on executing a timelock that has never been created', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+
+    await expect(timeLock.setNum(numHash, num)).to.be.revertedWithCustomError(
+      timeLock,
+      'InvalidTimeLockHash',
+    );
+  });
+
+  it('should revert on executing a timelock if buffer period has not passed', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+    expect(etaTimeLock).to.equal(eta);
+
+    await expect(timeLock.setNum(numHash, num)).to.be.revertedWithCustomError(
+      timeLock,
+      'TimeLockNotReady',
+    );
+  });
+
+  it('should execute timelock if buffer period has passed', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+    expect(etaTimeLock).to.equal(eta);
+
+    await network.provider.send('evm_increaseTime', [timeDelay]);
```

Comment:
Make this test hardhat agnostic, so we can run it on a live network. Use a smaller delay, 30s, for e.g. and if network is not hardhat, then sleep for that time. Can do this in a separate PR